corehq/apps/couch_sql_migration/asyncforms.py

```diff
@@ -91,20 +95,36 @@ def _try_to_process_form(self, wrapped_form, retries=0):
             return
         if self.queues.try_obj(case_ids, wrapped_form):
             self.pool.spawn(self._async_migrate_form, wrapped_form, case_ids)
-        elif self.queues.full:
-            gevent.sleep()  # swap greenlets
 
     @exit_on_error
     def _async_migrate_form(self, wrapped_form, case_ids):
         self.migrate_form(wrapped_form, case_ids)
         self.queues.release_lock(wrapped_form)
 
     def _try_to_empty_queues(self):
+        """Process forms waiting in the queue
+
+        All items in the queue will be processed if the queue becomes
+        full. This is done to ensure that no items become perpetually
+        stuck in the queue. This may be masking a bug in this class or
+        `PartiallyLockingQueue` since the theory of operation should
+        prevent starvation. In any case draining the queue periodically
+        is a good thing since there is a negative correlation between
+        the number of items in the queue and `queue.pop()` performance.
+        """
+        queue = self.queues
+        was_full = queue.full
         while True:
-            new_wrapped_form, case_ids = self.queues.pop()
-            if not new_wrapped_form:
+            form, case_ids = queue.pop()
+            if form is not None:
+                self.pool.spawn(self._async_migrate_form, form, case_ids)
+            elif was_full and queue:
+                assert queue.processing, "deadlock!"
+                wait_for_one_task_to_complete(self.pool)
```

Comment:
What's the idea behind waiting for just one task to complete (I assume that's one form?) as opposed to waiting for all of them to complete or not waiting for any of them?

Reply:
Yes, a "task" is a "form" in this context.

This only happens after attempting to pop the next available form from the queue, but instead of returning a form `queue.pop()` returned `None` meaning that next form(s) in the queue are waiting on other currently-being-processed forms to complete before they can start processing. This waits for one task since `queue.pop()` will always return `None` until at least one task has completed (and unlocked its case ids); it is possible that `queue.pop()` will _still_ return `None` after one task has completed, so it may need to wait again the next time around. At some point after waiting for enough (but not necessarily all) tasks to complete `queue.pop()` will return a form that can be processed.

Also of note: `_try_to_empty_queues` does not care if the tasks currently being processed have completed or not—it only cares that all forms that were waiting in the queue have been started if it `was_full` when it was called.